rust/lance/src/dataset/index/frag_reuse.rs

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use crate::dataset::transaction::{Operation, Transaction};
+use crate::index::frag_reuse::{build_frag_reuse_index_metadata, load_frag_reuse_index_details};
+use crate::Dataset;
+use lance_core::Error;
+use lance_index::frag_reuse::{FragReuseIndexDetails, FragReuseVersion, FRAG_REUSE_INDEX_NAME};
+use lance_index::DatasetIndexExt;
+use lance_table::format::Index;
+use log::warn;
+use roaring::RoaringBitmap;
+use snafu::location;
+
+/// Cleanup a fragment reuse index based on the current condition of the indices.
+/// If all the indices currently available are already caught up to as a specific reuse version,
+/// all older reuse versions (inclusive) can be cleaned up.
+/// 
+/// An index is considered caught up against a specific reuse version if
+/// 1. the index is created after or at the same dataset version as the reuse version
+/// 2. there is no old fragment in the version that is covered by the index and can be remapped.
+///    If an index's fragment bitmap is missing, we will consider it as caught up.
+///    Otherwise, we will never be able to clean up the reuse version.
+/// 
+/// Note that there could be a race condition that an index is being added during the cleanup,
+/// This will make that specific index not efficient until the next reindex,
+/// but it will not cause any correctness problem.
+pub async fn cleanup_frag_reuse_index(dataset: &mut Dataset) -> lance_core::Result<()> {
+    let indices = dataset.load_indices().await?;
+    let frag_reuse_index_meta = match indices.iter().find(|idx| idx.name == FRAG_REUSE_INDEX_NAME) {
+        None => return Ok(()),
+        Some(idx) => idx,
+    };
+
+    let frag_reuse_details = load_frag_reuse_index_details(dataset, frag_reuse_index_meta)
+        .await
+        .unwrap();
+
+    let mut retained_versions = Vec::new();
+    let mut fragment_bitmaps = RoaringBitmap::new();
+    for version in frag_reuse_details.versions.iter() {
+        if !indices.iter().all(|idx| is_index_remap_caught_up(version, idx).unwrap()) {
+            fragment_bitmaps.extend(
+                version
+                    .new_frags
+                    .iter()
+                    .map(|id| *id as u32)
+                    .collect::<Vec<_>>(),
+            );
+            retained_versions.push(version.clone());
+        }
+    }
+
+    let frag_reuse_index_details = FragReuseIndexDetails {
+        versions: retained_versions,
+    };
+
+    let new_index_meta =
+        build_frag_reuse_index_metadata(dataset, frag_reuse_index_details, fragment_bitmaps)
+            .await?;
+
+    let transaction = Transaction::new(
```

Comment:
note that this PR does not include the conflict resolution logic when there is a CreateIndex transaction vs a Rewrite transaction that both tries to create a new frag reuse index. That is handled in the next PR.